File: mycelium/phases/parsing.py
"""Phase 2: Tree-sitter AST to symbol extraction."""

from __future__ import annotations

import logging
import os

import tree_sitter

from mycelium.config import AnalysisConfig, SymbolType
from mycelium.graph.knowledge_graph import KnowledgeGraph
from mycelium.graph.namespace_index import NamespaceIndex
from mycelium.graph.symbol_table import SymbolTable
from mycelium.languages import get_analyser

logger = logging.getLogger(__name__)

# Cache parsers per language to avoid re-creating
_parsers: dict[str, tree_sitter.Parser] = {}


def _get_parser(analyser) -> tree_sitter.Parser | None:
    """Get or create a parser for the given analyser."""
    key = analyser.language_name
    if key not in _parsers:
        try:
            lang = analyser.get_language()
            _parsers[key] = tree_sitter.Parser(lang)
        except (RuntimeError, Exception) as e:
            logger.warning(f"Failed to initialise parser for {key}: {e}")
            return None
    return _parsers[key]


def run_parsing_phase(
    config: AnalysisConfig, kg: KnowledgeGraph, st: SymbolTable,
    ns_index: NamespaceIndex | None = None,
) -> None:
    """Parse source files and extract symbols."""
    _sym_counter = 0

    for file_data in kg.get_files():
        language = file_data.get("language")
        if not language:
            continue

        file_path = file_data["path"]
        ext = os.path.splitext(file_path)[1].lower()

        # Apply language filter
        if config.languages and language not in config.languages:
            continue

        analyser = get_analyser(ext)
        if analyser is None:
            continue

        # Check if analyser has is_available method (e.g., VB.NET)
        if hasattr(analyser, "is_available") and not analyser.is_available():
            continue

        parser = _get_parser(analyser)
        if parser is None:
            continue

        # Read file
        full_path = os.path.join(config.repo_path, file_path)
        try:
            with open(full_path, "rb") as f:
                source = f.read()
        except OSError as e:
            logger.warning(f"Failed to read {file_path}: {e}")
            continue

        # Parse
        try:
            tree = parser.parse(source)
        except Exception as e:
            logger.warning(f"Failed to parse {file_path}: {e}")
            continue

        # Extract symbols
        try:
            symbols = analyser.extract_symbols(tree, source, file_path)
        except Exception as e:
            logger.warning(f"Failed to extract symbols from {file_path}: {e}")
            continue

        # Assign IDs and add to graph + symbol table
        for symbol in symbols:
            _sym_counter += 1
            symbol.id = f"sym_{_sym_counter:04d}"
            symbol.language = language
            kg.add_symbol(symbol)
            st.add(symbol)

            # Register namespaces in the namespace index
            if ns_index and symbol.type == SymbolType.NAMESPACE:
                ns_index.register(symbol.name, file_path)

File: crates/mycelium-core/src/phases/parsing.rs
//! Phase 2: Tree-sitter AST parsing, extract symbols into SymbolTable + KnowledgeGraph + NamespaceIndex.

use std::collections::HashSet;
use std::path::Path;

use crate::config::AnalysisConfig;
use crate::graph::knowledge_graph::{KnowledgeGraph, NodeData};
use crate::graph::namespace_index::NamespaceIndex;
use crate::graph::symbol_table::SymbolTable;
use crate::languages::AnalyserRegistry;

/// Run the parsing phase: parse all source files and extract symbols.
pub fn run_parsing_phase(
    config: &AnalysisConfig,
    kg: &mut KnowledgeGraph,
    st: &mut SymbolTable,
    ns_index: &mut NamespaceIndex,
) {
    let registry = AnalyserRegistry::new();

    // Collect file paths from the knowledge graph
    let files: Vec<(String, Option<String>)> = kg
        .get_files()
        .into_iter()
        .filter_map(|n| {
            if let NodeData::File { path, language, .. } = n {
                Some((path.clone(), language.clone()))
            } else {
                None
            }
        })
        .collect();

    // Track used symbol IDs for deduplication
    let mut used_ids = HashSet::new();

    for (file_path, _language) in &files {
        let ext = Path::new(file_path)
            .extension()
            .map(|e| e.to_string_lossy().to_string())
            .unwrap_or_default();

        let analyser = match registry.get_by_extension(&ext) {
            Some(a) => a,
            None => continue,
        };

        let abs_path = Path::new(&config.repo_path).join(file_path);
        let source = match std::fs::read(&abs_path) {
            Ok(s) => s,
            Err(_) => continue,
        };

        // Parse with tree-sitter, using extension-specific language
        let mut parser = tree_sitter::Parser::new();
        let language = analyser.get_language_for_ext(&ext);
        if parser.set_language(&language).is_err() {
            continue;
        }
        let tree = match parser.parse(&source, None) {
            Some(t) => t,
            None => continue,
        };

        // Extract symbols
        let mut symbols = analyser.extract_symbols(&tree, &source, file_path);

        // Fix up symbol IDs: replace _pending_N with proper IDs
        for symbol in &mut symbols {
            let base_id = if let Some(ref parent) = symbol.parent {
                format!("{}:{}.{}", file_path, parent, symbol.name)
            } else {
                format!("{}:{}", file_path, symbol.name)
            };

            // Deduplicate IDs
            let mut id = base_id.clone();
            let mut counter = 1;
            while used_ids.contains(&id) {
                id = format!("{}_{}", base_id, counter);
                counter += 1;
            }
            used_ids.insert(id.clone());
            symbol.id = id;
        }

        for symbol in &symbols {
            kg.add_symbol(symbol);
            st.add(symbol);

            // Register namespaces
            if symbol.symbol_type.as_str() == "Namespace" {
                ns_index.register(&symbol.name, &symbol.file);
            }
        }
    }
}
